mesonbuild/mdevenv.py

```diff
@@ -0,0 +1,80 @@
+import os, subprocess
+import argparse
+import tempfile
+import shutil
+
+from pathlib import Path
+from . import build
+from .mesonlib import MesonException, is_windows
+
+import typing as T
+
+def add_arguments(parser: argparse.ArgumentParser) -> None:
+    parser.add_argument('-C', default='.', dest='wd',
+                        help='directory to cd into before running')
+    parser.add_argument('command', nargs=argparse.REMAINDER,
+                        help='Command to run in developer environment (default: interactive shell)')
+
+def get_windows_shell() -> str:
+    mesonbuild = Path(__file__).parent
+    script = mesonbuild / 'scripts' / 'cmd_or_ps.ps1'
+    command = ['powershell.exe', '-noprofile', '-executionpolicy', 'bypass', '-file', str(script)]
+    result = subprocess.check_output(command)
+    return result.decode().strip()
+
+def get_env(b: build.Build, build_dir: str) -> T.Dict[str, str]:
+    env = os.environ.copy()
+    for i in b.devenv:
+        env = i.get_env(env)
+
+    extra_env = build.EnvironmentVariables()
+    extra_env.set('MESON_DEVENV', ['1'])
+
+    meson_uninstalled = Path(build_dir) / 'meson-uninstalled'
+    if meson_uninstalled.is_dir():
+        extra_env.prepend('PKG_CONFIG_PATH', [str(meson_uninstalled)])
+
+    return extra_env.get_env(env)
+
+def run(options: argparse.Namespace) -> int:
+    options.wd = os.path.abspath(options.wd)
+    buildfile = Path(options.wd) / 'meson-private' / 'build.dat'
+    if not buildfile.is_file():
+        raise MesonException('Directory {!r} does not seem to be a Meson build directory.'.format(options.wd))
+    b = build.load(options.wd)
+
+    devenv = get_env(b, options.wd)
+
+    args = options.command
+    if not args:
+        prompt_prefix = '[{}]'.format(b.project_name)
+        if is_windows():
+            shell = get_windows_shell()
+            if shell == 'powershell.exe':
+                args = ['powershell.exe']
+                args += ['-NoLogo', '-NoExit']
+                prompt = 'function global:prompt {{  "{} PS " + $PWD + "> "}}'.format(prompt_prefix)
+                args += ['-Command', prompt]
+            else:
+                args = [os.environ.get("COMSPEC", r"C:\WINDOWS\system32\cmd.exe")]
+                args += ['/k', 'prompt {} $P$G'.format(prompt_prefix)]
+        else:
+            args = [os.environ.get("SHELL", os.path.realpath("/bin/sh"))]
+        if "bash" in args[0] and not os.environ.get("MESON_DISABLE_PS1_OVERRIDE"):
```

Comment:
I feel like this specialty stuff for UNIX-land is pointless and asking for pain. I can very easily change my prompt in my bashrc to do something special when `MESON_DEVENV` is defined. I already do something similar to Python virtual environments. As long as it is documented, developers can customize on their own. Plus it is going to become really ugly in this file when everyone in the world wants to add their random shell logic here.

Reply:
It's harder to do a different PS1 for each project in which you could be using devenv. IMHO it's useful to have that information that you're currently inside a devenv.

Reply:
The current copying of bashrc is completely painful, just create a new bashrc that does:

```
. ~/.bashrc
PS1="..."
```

I'm not terribly upset with the idea of doing this at all, though... it's basically like `venv/bin/activate` in python land.

Reply:
Then this PR should support a way to disable the default prompt behavior like Python virtual environments do.

```
[0 0 15:34:58] [tpartin@campise] [hse] [updates/NFSE-4592-followup3 $%=] [venv]
● $
```

This is my prompt and here `venv` is the name of my Python virtual environment which is very different than the default behavior.

Reply:
`export VIRTUAL_ENV_DISABLE_PROMPT=1` is what I do for Python.

Reply:
@eli-schwartz tbh this is a blind copy paste from gst-env.py, but I think you're right.

@tristan957 you can disable it with MESON_DISABLE_PS1_OVERRIDE, but that trick is missing from doc I think.

Reply:
oh sweet. I missed that in my ocde readthrough

Reply:
@eli-schwartz changed the rcfile as you suggested.
@tristan957 added  in the doc a mention about MESON_DISABLE_PS1_OVERRIDE.